qiskit_nature/operators/second_quantization/fermionic_op.py

```diff
@@ -464,6 +465,67 @@ def to_list(
             ]
         return self._to_dense_label_data()
 
+    def to_matrix(self, sparse: Optional[bool] = True) -> Union[csc_matrix, np.ndarray]:
+        """Convert to a matrix representation over the full fermionic Fock space in occupation number
+        basis. The basis states are ordered in increasing bitstring order as 0000, 0001, ..., 1111.
+
+        Args:
+            sparse: If true, the matrix is returned as a sparse csc_matrix, else it is returned as a
+            dense numpy array.
+
+        Returns:
+            The matrix of the operator in the Fock basis (scipy.sparse.csc_matrix or numpy.ndarray
+            with dtype=numpy.complex128)
+        """
+
+        csc_data, csc_col, csc_row = [], [], []
+
+        # loop over all columns of the matrix
+        for col_idx in range(2 ** self.register_length):
+            initial_occupations = [
+                occ == "1" for occ in f"{col_idx:0{self.register_length}b}"
+            ]
+            # loop over the terms in the operator data
+            for opstring, prefactor in self.reduce()._data:
+                # check if op string is the identity
+                if not opstring:
+                    csc_data.append(prefactor)
+                    csc_row.append(col_idx)
+                    csc_col.append(col_idx)
+                else:
+                    occupations = initial_occupations.copy()
+                    sign = 1
+                    mapped_to_zero = False
+
+                    # apply terms sequentially to the current basis state
+                    for label_primitive in reversed(opstring):
+                        occ = occupations[label_primitive.index]
+                        if label_primitive.is_creation == occ:
+                            # Applying the creation operator on an occupied state maps to zero. So
+                            # does applying the annihilation operator on an unoccupied state.
+                            mapped_to_zero = True
+                            break
+                        sign *= (-1) ** sum(occupations[: label_primitive.index])
+                        occupations[label_primitive.index] = not occ
+
+                    # add data point to matrix in the correct row
+                    if not mapped_to_zero:
+                        row_idx = int("".join([str(int(occ)) for occ in occupations]), 2)
```

Comment:
I think we can also improve this a bit. The back-and-forth conversion of `int` and `str` caught my eye a little bit.
Instead I propose the following:
```suggestion
                        row_idx = sum(int(occ) << idx for idx, occ in enumerate(occupations[::-1]))
```
This results in quite a significant speedup:
```python
>>> timeit.timeit("int(''.join([str(int(occ)) for occ in occs]), 2)", "occs = [True] * 32 + [False] * 32")
12.031811372999982
>>> timeit.timeit("sum(int(occ) << idx for idx, occ in enumerate(occs[::-1]))", "occs = [True] * 32 + [False] * 32")
7.646822761999829
```